symengine/symengine_rcp.h

```diff
@@ -146,8 +146,12 @@ template<class T2, class T1>
 inline RCP<T2> rcp_static_cast(const RCP<T1>& p1)
 {
     // Make the compiler check if the conversion is legal
+#if defined(TEUCHOS_DEBUG)
+    return rcp_dynamic_cast<T2>(p1);
+#else
     T2 *check = static_cast<T2*>(p1.get());
     return RCP<T2>(check);
+#endif
```

Comment:
You don't need to modify this file at all, because this file is only used in Release mode. The reason we use Teuchos in Debug mode is that it is actually quite difficult to catch all segfaults and invalid memory accesses, Teuchos does it right, and so we just use it. In Release mode we use our faster versions. So there is no point of adding Debug checks into our code, since it won't be perfect anyway and it's not used.